cmd/restic/cmd_cleanup.go

```diff
@@ -0,0 +1,372 @@
+package main
+
+import (
+	"golang.org/x/sync/errgroup"
+
+	"github.com/restic/restic/internal/crypto"
+	"github.com/restic/restic/internal/debug"
+	"github.com/restic/restic/internal/errors"
+	"github.com/restic/restic/internal/pack"
+	"github.com/restic/restic/internal/repository"
+	"github.com/restic/restic/internal/restic"
+
+	"github.com/spf13/cobra"
+)
+
+var cmdCleanup = &cobra.Command{
+	Use:   "cleanup [flags]",
+	Short: "Cleanup unused data",
+	Long: `
+The "cleanup" cleans up data in index and pack files
+that is not referenced in any snapshot file.
+
+When calling this command without flags, only packs
+that are completely unused are deleted and index files are cleaned.
+You can specify additional conditions to repack
+packs that are only partly used or too small.
+These packs will be downloaded and uploaded again which can be
+quite time-consuming for remote repositories.
+`,
+	DisableAutoGenTag: true,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return runCleanup(cleanupOptions, globalOptions)
+	},
+}
+
+// CleanupIndexOptions collects all options for the cleanup command.
+type CleanupOptions struct {
+	DryRun            bool
+	DataUnusedPercent int8
+	DataUnusedSpace   int64
+	DataUsedSpace     int64
+	TreeUnusedPercent int8
+	TreeUnusedSpace   int64
+	TreeUsedSpace     int64
+	RepackMixed       bool
+}
+
+var cleanupOptions CleanupOptions
+
+func init() {
+	cmdRoot.AddCommand(cmdCleanup)
+
+	f := cmdCleanup.Flags()
+	f.BoolVarP(&cleanupOptions.DryRun, "dry-run", "n", false, "do not delete anything, just print what would be done")
+	f.Int8Var(&cleanupOptions.DataUnusedPercent, "data-unused-percent", -1, "if > 0, repack data packs with >= given % unused space")
+	f.Int64Var(&cleanupOptions.DataUnusedSpace, "data-unused-space", -1, "if > 0, repack data packs with >= given bytes unused space")
+	f.Int64Var(&cleanupOptions.DataUsedSpace, "data-used-space", -1, "repack data packs with <= given bytes used space")
```

Comment:
You are right. I'll change this and actually like your suggestions of flags. However, about `max-unused-percentage` I would even change this to a repository-wide flag instead of a pack-specific flag. In fact, a user would not care if there are single packs that have too much unused space, but would care if there is too much unused repository space. Treating this as a  parameter for the whole repository gives much more possibilities to optimize what to keep, what to delete and what to repack - while keeping and deleting are the preferred options.

I also do have the feeling that treating duplicate blobs correctly (which is also needed to treat aborted prunes correctly where index files have already been written) needs a "global" optimization strategy and then a repository-wide parameter enables much better solutions.

Reply:
A repository wide flag sounds good. The pack selection could look like this: Sort all packs by their used space percentage (lowest to highest) and then select enough packs with low usage to satisfy the space overhead threshold.

The simplest solution for duplicates would be to just rewrite all packs containing a duplicate blob. Or to count each duplicate blob as unused for the usage threshold (but of course not blindly delete its pack).

Reply:
This is what I have already started to implement :+1: 

The correct and - especially - optimal handling of duplicate blobs is really nasty. E.g. if there is a pack full of duplicates it could be simply deleted. Of course only if there is no repository damage which could be healed by this specific pack. And so on...

I also realized that programming this out in a complete way is not worth the effort as duplicate blobs should always be an edge case that will be sanitized by a prune.
Hence I added a flag `--repack-duplicates` (default=true) which repacks all duplicates.